fedora_messaging/twisted/protocol.py

```diff
@@ -301,88 +286,152 @@ def publish(self, message, exchange):
-        deferred.addErrback(
-            lambda f: _log.failure, "_read failed on consumer {c}", c=consumer
-        )
-        _log.info("Successfully registered AMQP consumer {c}", c=consumer)
+            This can happen, for example, when the queue is deleted.
+            To handle this, we do the necessary book-keeping to remove the consumer
+            and then fire the errback on the consumer so the caller of
+            :func:`fedora_messaging.api.consume` can decide what to do.
+
+            Args:
+                frame (pika.frame.Method): The cancel method from the server,
+                    unused here because we already know what consumer is being
+                    canceled.
+            """
+            _std_log.error("%r was canceled by the AMQP broker!", consumer)
+
+            # If client and server are racing to cancel it might already be gone which
+            # is why both are marked as no cover.
+            try:
+                del self._consumers[consumer.queue]
+            except KeyError:  # pragma: no cover
+                pass
+            try:
+                del self.factory._consumers[consumer.queue]
+            except KeyError:  # pragma: no cover
+                pass
+            consumer._running = False
+            consumer.result.errback(fail=ConsumerCancelled())
+
+        try:
+            consumer._channel.add_on_cancel_callback(on_cancel_callback)
+        except AttributeError:
+            pass  # pika 1.0.0+
+
+        def read_loop_errback(failure):
+            """
+            Handle errors coming out of the read loop.
+
+            There are two basic categories of errors: ones where the ``consumer.result``
+            Deferred needs to be fired because the error is not recoverable, ones
+            where we can recover from by letting the connection restart, and ones
+            which are fatal for this consumer only (the queue was deleted by an
+            administrator).
+
+            Args:
+                failure (twisted.python.failure.Failure): The exception raised by
+                    the read loop encapsulated in a Failure.
+            """
+            if failure.check(
+                pika.exceptions.ChannelClosed,
+                error.ConnectionDone,
+                error.ConnectionLost,
+            ):
+                _std_log.warn(
+                    "Consumer halted (%r) while waiting for a message; "
+                    "the connection should restart.",
+                    failure,
+                )
+            elif failure.check(pika.exceptions.ConsumerCancelled):
+                # Pika 1.0.0+ raises this exception. To support previous versions
+                # we register a callback (called below) ourselves with the channel.
+                on_cancel_callback(None)
+            else:
+                consumer.result.errback(failure)
+                self.cancel(consumer)
+
+        consumer._read_loop = self._read(queue_object, consumer)
```

Comment:
The read loop is an internal implementation detail I didn't want to expose to users of the API. The Consumer object is mostly not defined in this module because there's already a Consumer namedtuple (which I couldn't use because it's immutable) and naming got weird when combined with autodocs, but it's definitely meant to be something produced and consumed by the protocol, and there are internal bits users don't need to know about.